### ISSUE TITLE
Decode{Int8, Int16, Int32} must check if their values are less than minimum

### ISSUE THREAD

**Issue by odeke-em** [tendermint/go-amino#120]
The code in 
https://github.com/tendermint/go-wire/blob/df56f863670d5654f1314b20572d58dd5bf8e548/decoder.go#L12-L51

does a great job of checking that the values aren't greater than their respective max values i.e.:
* math.MaxInt8
* math.MaxInt16
* math.MaxInt32

However, they don't check if the values are less than the respective minimums:
* math.MinInt8
* math.MinInt16
* math.MinInt32

We need to fix this as potentially this is a way for data to be corrupted and an attack vector because I can sneak in say -0xFF to be decoded by int8 as instead 1, and -0x100 as 0 for int8 too -- data corruption!

The same applies for every one of those types, just push it to a casting boundary of
-(math.MaxU* + 1), -(math.Max* + 2) to give you 1 and 0 respectively
as per https://play.golang.org/p/ACtLH18S0AZ

**Comment by odeke-em** [comment]
Sample runnable test patch

```diff
diff --git a/codec_test.go b/codec_test.go
index fcab311..1286414 100644
--- a/codec_test.go
+++ b/codec_test.go
@@ -1,6 +1,8 @@
 package wire_test
 
 import (
+	"encoding/binary"
+	"strings"
 	"testing"
 	"time"
 
@@ -93,3 +95,50 @@ func TestMarshalUnmarshalBinaryPointer4(t *testing.T) {
 	assert.Equal(t, s, *s2)
 
 }
+
+func TestDecodeInt8(t *testing.T) {
+	// DecodeInt8 uses binary.Varint so we need to make
+	// sure that all the values out of the range of [-128, 127]
+	// return an error.
+	tests := []struct {
+		in      int64
+		wantErr string
+		want    int8
+	}{
+		{in: 0x7F, want: 0x7F},
+		{in: -0x7F, want: -0x7F},
+		{in: -0x80, want: -0x80},
+		{in: 0x10, want: 0x10},
+
+		{in: -0xFF, wantErr: "decoding int8"},
+		{in: 0xFF, wantErr: "decoding int8"},
+		{in: 0x100, wantErr: "decoding int8"},
+		{in: -0x100, wantErr: "decoding int8"},
+	}
+
+	buf := make([]byte, 10)
+	for i, tt := range tests {
+		n := binary.PutVarint(buf, tt.in)
+		gotI8, gotN, err := wire.DecodeInt8(buf[:n])
+		if tt.wantErr != "" {
+			if err == nil {
+				t.Errorf("#%d expected error=%q", i, tt.wantErr)
+			} else if !strings.Contains(err.Error(), tt.wantErr) {
+				t.Errorf("#%d\ngotErr=%q\nwantSegment=%q", i, err, tt.wantErr)
+			}
+			continue
+		}
+
+		if err != nil {
+			t.Errorf("#%d unexpected error: %v", i, err)
+			continue
+		}
+
+		if wantI8 := tt.want; gotI8 != wantI8 {
+			t.Errorf("#%d gotI8=%d wantI8=%d", i, gotI8, wantI8)
+		}
+		if wantN := n; gotN != wantN {
+			t.Errorf("#%d gotN=%d wantN=%d", i, gotN, wantN)
+		}
+	}
+}
```

and then a fix
```diff
diff --git a/decoder.go b/decoder.go
index 2acc79a..30a5565 100644
--- a/decoder.go
+++ b/decoder.go
@@ -17,7 +17,7 @@ func DecodeInt8(bz []byte) (i int8, n int, err error) {
 	if err != nil {
 		return
 	}
-	if i64 > int64(math.MaxInt8) {
+	if i64 < int64(math.MinInt8) || i64 > int64(math.MaxInt8) {
 		err = errors.New("EOF decoding int8")
 		return
 	}
```

**Comment by jaekwon** [comment]
Sounds good, can you please make a PR?

**Comment by odeke-em** [comment]
For sure, let me do so.

**Comment by liamsi** [comment]
resolved via #125
